File: paper/src/main/java/com/gitlab/aecsocket/sokol/paper/SokolCommand.java
package com.gitlab.aecsocket.sokol.paper;

import cloud.commandframework.ArgumentDescription;
import cloud.commandframework.arguments.flags.CommandFlag;
import cloud.commandframework.arguments.standard.EnumArgument;
import cloud.commandframework.arguments.standard.IntegerArgument;
import cloud.commandframework.arguments.standard.StringArgument;
import cloud.commandframework.bukkit.parsers.PlayerArgument;
import cloud.commandframework.bukkit.parsers.selector.MultiplePlayerSelectorArgument;
import cloud.commandframework.context.CommandContext;
import com.gitlab.aecsocket.minecommons.core.CollectionBuilder;
import com.gitlab.aecsocket.minecommons.core.Components;
import com.gitlab.aecsocket.minecommons.paper.plugin.BaseCommand;
import com.gitlab.aecsocket.sokol.core.node.ItemCreationException;
import com.gitlab.aecsocket.sokol.core.node.NodePath;
import com.gitlab.aecsocket.sokol.core.registry.Keyed;
import com.gitlab.aecsocket.sokol.core.stat.Stat;
import com.gitlab.aecsocket.sokol.core.stat.StatIntermediate;
import com.gitlab.aecsocket.sokol.core.stat.StatMap;
import com.gitlab.aecsocket.sokol.paper.command.BlueprintArgument;
import com.gitlab.aecsocket.sokol.paper.command.ComponentArgument;
import com.gitlab.aecsocket.sokol.paper.command.NodeArgument;
import com.gitlab.aecsocket.sokol.paper.impl.*;
import com.gitlab.aecsocket.sokol.paper.wrapper.user.PlayerUser;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.JoinConfiguration;
import net.kyori.adventure.text.event.ClickEvent;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.*;
import java.util.stream.Collectors;

import static net.kyori.adventure.text.Component.*;
import static net.kyori.adventure.text.JoinConfiguration.*;

/* package */ final class SokolCommand extends BaseCommand<SokolPlugin> {
    private static final JoinConfiguration separator = separator(text(", ", NamedTextColor.GRAY));
    private static final List<String> registryTypes = List.of(
            PaperComponent.class.getSimpleName(),
            PaperBlueprint.class.getSimpleName()
    );

    SokolCommand(SokolPlugin plugin) throws Exception {
        super(plugin, "sokol",
                (mgr, root) -> mgr.commandBuilder(root, ArgumentDescription.of("Plugin main command.")));

        manager.command(root
                .literal("list", ArgumentDescription.of("Lists all registered objects."))
                .flag(CommandFlag.newBuilder("type")
                        .withAliases("t")
                        .withArgument(StringArgument.<CommandSender>newBuilder("value")
                                .quoted()
                                .withSuggestionsProvider((ctx, inp) -> registryTypes))
                        .withDescription(ArgumentDescription.of("The type of object to get.")))
                .flag(CommandFlag.newBuilder("filter")
                        .withAliases("f")
                        .withArgument(StringArgument.newBuilder("value")
                                .quoted()
                                .withSuggestionsProvider((ctx, inp) -> CollectionBuilder.list(new ArrayList<>(plugin.components().keySet()))
                                        .add(plugin.blueprints().keySet())
                                        .get()))
                        .withDescription(ArgumentDescription.of("The filter for the name/ID.")))
                .permission("%s.command.list".formatted(rootName))
                .handler(c -> handle(c, this::list)));
        manager.command(root
                .literal("tree", ArgumentDescription.of("Shows a node tree of an equipped item."))
                .argument(EnumArgument.optional(EquipmentSlot.class, "slot"), ArgumentDescription.of("The slot to get the item in."))
                .argument(PlayerArgument.optional("target"), ArgumentDescription.of("The player to get the item from."))
                .flag(CommandFlag.newBuilder("contextless")
                        .withAliases("c")
                        .withDescription(ArgumentDescription.of("Removes the context of you being the user, when generating info.")))
                .permission("%s.command.tree".formatted(rootName))
                .handler(c -> handle(c, this::tree)));
        manager.command(root
                .literal("component", ArgumentDescription.of("Outputs detailed information on a registered component."))
                .argument(ComponentArgument.of(plugin, "component"), ArgumentDescription.of("The component to list information on."))
                .permission("%s.command.component".formatted(rootName))
                .handler(c -> handle(c, this::component)));
        manager.command(root
                .literal("blueprint", ArgumentDescription.of("Outputs detailed information on a registered blueprint."))
                .argument(BlueprintArgument.of(plugin, "blueprint"), ArgumentDescription.of("The blueprint to list information on."))
                .permission("%s.command.blueprint".formatted(rootName))
                .handler(c -> handle(c, this::blueprint)));
        manager.command(root
                .literal("give", ArgumentDescription.of("Gives a component/node tree to specified players."))
                .argument(MultiplePlayerSelectorArgument.of("targets"), ArgumentDescription.of("The players to give the item to."))
                .argument(NodeArgument.of(plugin, "node"), ArgumentDescription.of("The component/node tree to give."))
                .argument(IntegerArgument.<CommandSender>newBuilder("amount").withMin(1).asOptional(), ArgumentDescription.of("The amount of the item to give."))
                .permission("%s.command.give".formatted(rootName))
                .handler(c -> handle(c, this::give)));
        manager.command(root
                .literal("build", ArgumentDescription.of("Gives a built blueprint to specified players."))
                .argument(MultiplePlayerSelectorArgument.of("targets"), ArgumentDescription.of("The players to give the item to."))
                .argument(BlueprintArgument.of(plugin, "blueprint"), ArgumentDescription.of("The blueprint to give."))
                .argument(IntegerArgument.<CommandSender>newBuilder("amount").withMin(1).asOptional(), ArgumentDescription.of("The amount of the item to give."))
                .permission("%s.command.build".formatted(rootName))
                .handler(c -> handle(c, this::build)));
        // todo list command
    }

    private List<Component> formatStats(Locale locale, StatMap map) {
        List<Component> lines = new ArrayList<>();
        lc.lines(locale, PREFIX_COMMAND + ".stats.header",
                        "amount", map.size()+"")
                .ifPresent(lines::addAll);
        for (var entry : map.entrySet()) {
            String key = entry.getKey();
            Stat.Node<?> node = entry.getValue();
            List<? extends Stat.Node<?>> chain = node.asList();
            lc.lines(locale, PREFIX_COMMAND + ".stats.entry",
                            "name", node.stat().render(locale, lc),
                            "key", key,
                            "nodes", chain.size() + "",
                            "chain", join(separator, chain.stream().map(n -> n.value().render(locale, lc)).collect(Collectors.toList())))
                    .ifPresent(lines::addAll);
        }

        return lines;
    }

    private void list(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        String type = ctx.flags().get("type");
        if (type != null) type = type.toLowerCase(Locale.ROOT);
        String filter = ctx.flags().get("filter");
        if (filter != null) filter = filter.toLowerCase(Locale.ROOT);

        List<Keyed> registered = new ArrayList<>(plugin.components().values());
        registered.addAll(plugin.blueprints().values());
        int results = 0;
        for (var object : registered) {
            Component rendered = object.render(locale, lc);
            String id = object.id();
            if (type != null) {
                String typeName = object.getClass().getName().toLowerCase(Locale.ROOT);
                if (!typeName.contains(type))
                    continue;
            }
            if (filter != null) {
                String testId = id.toLowerCase(Locale.ROOT);
                String name = PlainTextComponentSerializer.plainText().serialize(rendered).toLowerCase(Locale.ROOT);
                if (!testId.contains(filter) && !name.contains(filter))
                    continue;
            }

            ++results;
            String command = "/" + rootName + " " + (
                    object instanceof PaperComponent ? "component"
                    : object instanceof PaperBlueprint ? "blueprint"
                    : "?") + " " + object.id();
            Component hover = fullInfoHover(locale, command);
            ClickEvent click = ClickEvent.runCommand(command);
            lc.lines(locale, PREFIX_COMMAND + ".list.entry",
                    "type", object.getClass().getSimpleName(),
                    "name", rendered,
                    "id", id)
                    .ifPresent(m -> m.forEach(c -> sender.sendMessage(c.hoverEvent(hover).clickEvent(click))));
        }
        send(sender, locale, "list.total",
                "results", ""+results);
    }

    private Component constant(Locale locale, boolean value) {
        return lc.safe(locale, "constant." + value);
    }

    private Component fullInfoHover(Locale locale, String command) {
        return lc.lines(locale, PREFIX_COMMAND + ".full_info_hover",
                        "command", command)
                .map(m -> join(separator(newline()), m))
                .orElse(empty());
    }

    private void tree0(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender, Component indent, int depth, PaperNode node) {
        for (var entry : node.nodes().entrySet()) {
            String key = entry.getKey();
            PaperNode child = entry.getValue();
            PaperComponent component = child.value();

            String command = "/" + rootName + " component " + component.id();
            Component hover = fullInfoHover(locale, command);
            ClickEvent click = ClickEvent.runCommand(command);
            lc.lines(locale, PREFIX_COMMAND + ".tree.node",
                            "indent", Components.repeat(indent, depth),
                            "slot", node.value().slot(key).orElseThrow(IllegalStateException::new).render(locale, lc),
                            "key", key,
                            "name", component.render(locale, lc),
                            "id", component.id())
                    .ifPresent(m -> m.forEach(c -> sender.sendMessage(c.hoverEvent(hover).clickEvent(click))));
            tree0(ctx, sender, locale, pSender, indent, depth + 1, child);
        }
    }

    private void tree(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender, PaperNode node, String type) {
        String command = "/" + rootName + " component " + node.value().id();
        Component hover = fullInfoHover(locale, command);
        ClickEvent click = ClickEvent.runCommand(command);
        lc.lines(locale, PREFIX_COMMAND + ".tree." + type,
                        "type", node.value().getClass().getSimpleName(),
                        "name", node.value().render(locale, lc),
                        "id", node.value().id())
                .ifPresent(m -> m.forEach(c -> sender.sendMessage(c.hoverEvent(hover).clickEvent(click))));
        tree0(ctx, sender, locale, pSender, lc.safe(locale, PREFIX_COMMAND + ".tree.indent"), 0, node);
    }

    private void tree(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        EquipmentSlot slot = ctx.getOrDefault("slot", EquipmentSlot.HAND);
        Player target = defaultedArg(ctx, "target", pSender, p -> p);

        //noinspection ConstantConditions
        ItemStack item = target.getInventory().getItem(slot);
        PaperNode node = plugin.persistence().safeLoad(item)
                .orElseThrow(() -> error("item_not_tree"));
        if (pSender == null || ctx.flags().isPresent("contextless"))
            node.initialize(locale);
        else
            node.initialize(PlayerUser.user(plugin, pSender));
        tree(ctx, sender, locale, pSender, node, "header");

        node.treeData().ifPresent(treeData -> {
            List<NodePath> incomplete = treeData.incomplete();
            if (incomplete.isEmpty())
                send(sender, locale, "tree.complete.complete");
            else {
                send(sender, locale, "tree.complete.incomplete",
                        "missing", ""+incomplete.size());
                for (var path : incomplete)
                    send(sender, locale, "tree.complete.incomplete_entry",
                            "path", ""+path);
            }
            StatMap stats = treeData.stats();
            Component hover = join(separator(newline()), formatStats(locale, stats));
            lc.lines(locale, PREFIX_COMMAND + ".tree.stats",
                    "amount", ""+stats.size())
                    .ifPresent(m -> m.forEach(c -> sender.sendMessage(c.hoverEvent(hover))));
        });
    }

    private void component(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        PaperComponent component = ctx.get("component");

        send(sender, locale, "tree.header",
                "type", component.getClass().getSimpleName(),
                "name", component.render(locale, lc),
                "id", component.id());

        component.renderDescription(locale, lc).ifPresent(desc -> {
            for (var line : desc)
                send(sender, locale, "component.description",
                        "line", line);
        });
        send(sender, locale, "component.tags",
                "tags", String.join(", ", component.tags()));

        send(sender, locale, "component.slots",
                "amount", component.slots().size()+"");
        for (var entry : component.slots().entrySet()) {
            String key = entry.getKey();
            PaperSlot slot = entry.getValue();

            Component hover = slot.rule().render(locale, lc);
            lc.lines(locale, PREFIX_COMMAND + ".component.slot",
                    "name", slot.render(locale, lc),
                    "key", key,
                    "tags", String.join(", ", slot.tags()),
                    "offset", slot.offset()+"")
                    .ifPresent(m -> m.forEach(c -> sender.sendMessage(c.hoverEvent(hover))));
        }

        send(sender, locale, "component.features",
                "amount", component.features().size()+"");
        for (var entry : component.features().entrySet()) {
            String id = entry.getKey();
            PaperFeature<?> feature = entry.getValue();

            Component hover = feature.renderConfig(locale, lc)
                    .map(cfg -> join(separator(newline()), cfg))
                    .orElse(empty());
            lc.lines(locale, PREFIX_COMMAND + ".component.feature",
                    "name", feature.render(locale, lc),
                    "id", id,
                    "description", feature.renderDescription(locale, lc))
                    .ifPresent(m -> m.forEach(c -> sender.sendMessage(c.hoverEvent(hover))));
        }

        List<StatIntermediate.MapData> stats = component.stats().join();
        send(sender, locale, "component.stats",
                "amount", stats.size()+"");
        for (var data : stats) {
            StatMap map = data.stats();
            List<Component> lines = formatStats(locale, data.stats());
            Component hover = join(separator(newline()), lines);
            lc.lines(locale, PREFIX_COMMAND + ".component.stat_map",
                    "priority", data.priority().toString(),
                    "amount", map.size()+"",
                    "rule", data.rule().render(locale, lc))
                    .ifPresent(m -> m.forEach(c -> sender.sendMessage(c.hoverEvent(hover))));
        }
    }

    private void blueprint(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        PaperBlueprint blueprint = ctx.get("blueprint");

        send(sender, locale, "tree.header",
                "type", blueprint.getClass().getSimpleName(),
                "name", blueprint.render(locale, lc),
                "id", blueprint.id());
        blueprint.renderDescription(locale, lc).ifPresent(desc -> {
            for (var line : desc)
                send(sender, locale, "tree.description",
                        "line", line);
        });

        PaperNode node = blueprint.node();
        tree(ctx, sender, locale, pSender, node, "root");
    }

    private void give(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender, PaperNode node) {
        // Make sure that the item can actually be built
        ItemStack baseItem;
        try {
            if (pSender == null)
                baseItem = node.createItem(locale).handle();
            else
                baseItem = node.createItem(PlayerUser.user(plugin, pSender)).handle();
        } catch (ItemCreationException e) {
            throw error("item_creation", e);
        }
        //noinspection ConstantConditions
        int amount = ctx.getOrDefault("amount", 1);
        List<Player> targets = targets(ctx, "targets", pSender);

        for (Player target : targets) {
            ItemStack item;
            // We shouldn't *need* this try/catch, but just in case...
            try {
                item = node.createItem(PlayerUser.user(plugin, target)).handle();
            } catch (ItemCreationException e) {
                throw error("item_creation", e);
            }
            Inventory inventory = target.getInventory();
            for (int i = 0; i < amount; i++)
                inventory.addItem(item);
        }

        send(sender, locale, "give",
                "amount", amount+"",
                "item", baseItem.displayName(),
                "target", targets.size() == 1
                        ? targets.get(0).displayName()
                        : targets.size()+"");
    }

    private void give(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        give(ctx, sender, locale, pSender, ctx.get("node"));
    }

    private void build(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        give(ctx, sender, locale, pSender, ctx.<PaperBlueprint>get("blueprint").create());
    }
}

File: paper/src/main/kotlin/com/gitlab/aecsocket/sokol/paper/SokolCommand.kt
package com.gitlab.aecsocket.sokol.paper

import cloud.commandframework.arguments.standard.IntegerArgument
import cloud.commandframework.bukkit.parsers.location.LocationArgument
import cloud.commandframework.bukkit.parsers.selector.MultiplePlayerSelectorArgument
import com.gitlab.aecsocket.alexandria.core.extension.value
import com.gitlab.aecsocket.alexandria.paper.BaseCommand
import com.gitlab.aecsocket.alexandria.paper.Context
import com.gitlab.aecsocket.alexandria.paper.desc
import com.gitlab.aecsocket.craftbullet.core.Timings
import com.gitlab.aecsocket.glossa.core.I18N
import com.gitlab.aecsocket.sokol.core.*
import com.gitlab.aecsocket.sokol.paper.component.ItemHolder
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.Component.text
import org.bukkit.Location
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

private val TIMING_INTERVALS = listOf(5, 30, 60)

internal class SokolCommand(
    override val plugin: Sokol
) : BaseCommand(plugin) {
    init {
        manager.command(root
            .literal("stats", desc("Shows info on how the plugin is running."))
            .permission(perm("stats"))
            .handler { handle(it, ::stats) })
        manager.command(root
            .literal("systems", desc("Shows registered system information."))
            .permission(perm("systems"))
            .handler { handle(it, ::systems) })

        manager.command(root
            .literal("give", desc("Gives an entity item to a player."))
            .argument(MultiplePlayerSelectorArgument.of("targets"), desc("Players to give to."))
            .argument(IntegerArgument.newBuilder<CommandSender>("amount")
                .withMin(1), desc("Amount of items to give."))
            .argument(EntityArgument(plugin, "entity"), desc("Entity to give."))
            .permission(perm("give"))
            .handler { handle(it, ::give) })
        manager.command(root
            .literal("summon", desc("Summons an entity mob."))
            .argument(LocationArgument.of("location"), desc("Where to spawn the mob."))
            .argument(IntegerArgument.newBuilder<CommandSender>("amount")
                .withMin(1), desc("Amount of mobs to spawn."))
            .argument(EntityArgument(plugin, "entity"), desc("Entity to spawn."))
            .permission(perm("summon"))
            .handler { handle(it, ::summon) })
    }

    private lateinit var mProfiled: ComponentMapper<Profiled>
    private lateinit var mItemHolder: ComponentMapper<ItemHolder>

    internal fun enable() {
        mProfiled = plugin.engine.mapper()
        mItemHolder = plugin.engine.mapper()
    }

    fun stats(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        fun sendTimings(baseTimings: Timings, headerKey: String) {
            plugin.sendMessage(sender, i18n.csafe(headerKey) {
                icu("time_last", baseTimings.allEntries().lastOrNull() ?: "?")
            })

            TIMING_INTERVALS.forEach { interval ->
                val intervalMs = interval * 1000L
                val timings = baseTimings.lastEntries(intervalMs)
                plugin.sendMessage(sender, i18n.csafe("stats.timing") {
                    icu("interval_ms", intervalMs)
                    icu("interval_sec", intervalMs / 1000.0)
                    icu("time_avg", timings.average())
                    icu("time_min", timings.min())
                    icu("time_max", timings.max())
                })
            }
        }

        sendTimings(plugin.timings, "stats.timings")

        plugin.sendMessage(sender, i18n.csafe("stats.object_types"))

        plugin.resolver.lastStats.forEach { (type, stats) ->
            val (candidates, updated) = stats
            val percent = updated.toDouble() / candidates
            plugin.sendMessage(sender, i18n.csafe("stats.object_type") {
                icu("type", type.key)
                icu("candidates", candidates)
                icu("updated", updated)
                icu("percent", percent)
            })
        }
    }

    fun systems(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val systemTypes = plugin.engine.systems()
            .mapNotNull { it::class.simpleName }

        plugin.sendMessage(sender, i18n.csafe("systems.header"))
        systemTypes.forEachIndexed { idx, type ->
            plugin.sendMessage(sender, i18n.csafe("systems.line") {
                icu("index", idx + 1)
                icu("type", type)
            })
        }
    }

    fun give(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val targets = ctx.players("targets", sender, i18n)
        val amount = ctx.value("amount") { 1 }
        val entity = ctx.get<SokolEntity>("entity")

        targets.forEach { target ->
            val targetEntity = entity.copyOf()
            mItemHolder.set(targetEntity, ItemHolder.byMob(target))
            val item = plugin.hoster.hostItem(targetEntity)
            item.amount = amount
            target.inventory.addItem(item)
        }

        if (sender is Player)
            mItemHolder.set(entity, ItemHolder.byMob(sender))
        val item = plugin.hoster.hostItem(entity)

        plugin.sendMessage(sender, i18n.csafe("give") {
            subst("item", item.displayName())
            subst("target", if (targets.size == 1) targets.first().name() else text(targets.size))
            icu("amount", amount)
        })
    }

    fun summon(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val location = ctx.get<Location>("location")
        val amount = ctx.value("amount") { 1 }
        val entity = ctx.get<SokolEntity>("entity")

        repeat(amount) {
            plugin.hoster.hostMob(entity.copyOf(), location)
        }

        val profile = mProfiled.get(entity).profile

        plugin.sendMessage(sender, i18n.csafe("summon") {
            subst("id", text(profile.id))
            icu("amount", amount)
        })
    }

    /*init {
        manager.command(root
            .literal("stats", desc("Show stats for the object resolver."))
            .permission(perm("stats"))
            .handler { handle(it, ::stats) })
        manager.command(root
            .literal("systems", desc("See the system information."))
            .permission(perm("systems"))
            .handler { handle(it, ::systems) })
        manager.command(root
            .literal("give", desc("Creates and gives an item blueprint to a player."))
            .argument(MultiplePlayerSelectorArgument.of("targets"), desc("Players to give to."))
            .argument(IntegerArgument.newBuilder<CommandSender>("amount")
                .withMin(1), desc("Amount of items to give."))
            .argument(KeyedEntityBlueprintArgument(plugin, "blueprint"), desc("The blueprint to use."))
            .permission(perm("give"))
            .handler { handle(it, ::give) })
        manager.command(root
            .literal("summon", desc("Creates and summons a mob blueprint."))
            .argument(LocationArgument.of("location"), desc("Where to spawn the mob."))
            .argument(IntegerArgument.newBuilder<CommandSender>("amount")
                .withMin(1), desc("The amount of mobs to spawn."))
            .argument(KeyedEntityBlueprintArgument(plugin, "blueprint"), desc("The blueprint to use."))
            .permission(perm("summon"))
            .handler { handle(it, ::summon) })

        val holding = root
            .literal("holding", desc("Tools for manipulating the current held item."))
        manager.command(holding
            .literal("freeze", desc("Freeze the held item in place."))
            .argument(BooleanArgument.optional("enable"), desc("If the item should be frozen."))
            .permission(perm("holding.freeze"))
            .senderType(Player::class.java)
            .handler { handle(it, ::holdingFreeze) })
        manager.command(holding
            .literal("shape", desc("Draw the shape from a config on the held item."))
            .argument(ConfigurationNodeArgument("config", { AlexandriaAPI.configLoader().buildAndLoadString(it) }), desc("The configuration node."))
            .permission(perm("holding.shape"))
            .senderType(Player::class.java)
            .handler { handle(it, ::holdingShape) })
        manager.command(holding
            .literal("slot-shapes", desc("Draw the shapes of all slots on the held item."))
            .argument(BooleanArgument.optional("enable"), desc("If the slots should be drawn."))
            .permission(perm("holding.slot-shapes"))
            .senderType(Player::class.java)
            .handler { handle(it, ::holdingSlotShapes) })

        val state = root
            .literal("state", desc("Access the state of an entity in the world."))
        val stateRead = state
            .literal("read", desc("Read the state of an entity in the world."))
        manager.command(stateRead
            .literal("mob", desc("Read from a mob."))
            .argument(MultipleEntitySelectorArgument.of("targets"), desc("Mobs to read entities from."))
            .permission(perm("state.read.mob"))
            .handler { handle(it, ::stateReadMob) })
        manager.command(stateRead
            .literal("item", desc("Read from an item in a player inventory."))
            .argument(PlayerInventorySlotArgument("slot", required = false), desc("Slot to read item from."))
            .argument(MultiplePlayerSelectorArgument.optional("targets"), desc("Players to get items from."))
            .permission(perm("state.read.item"))
            .handler { handle(it, ::stateReadItem) })
        val stateWrite = state
            .literal("write", desc(""))
        manager.command(stateWrite
            .literal("mob", desc(""))
            .argument(MultipleEntitySelectorArgument.of("targets"), desc(""))
            .argument(ConfigurationNodeArgument("data", { AlexandriaAPI.configLoader().buildAndLoadString(it) }), desc(""))
            .permission(perm("state.write.mob"))
            .handler { handle(it, ::stateWriteMob) })

        val composite = root
            .literal("composite", desc("Read the info of a composite tree of an entity."))
        manager.command(composite
            .literal("mob", desc("Read from a mob."))
            .argument(MultipleEntitySelectorArgument.of("targets"), desc("Mobs to read entities from."))
            .permission(perm("composite.mob"))
            .handler { handle(it, ::compositeReadMob) })
        manager.command(composite
            .literal("item", desc("Read from an item in a player inventory."))
            .argument(PlayerInventorySlotArgument("slot", required = false), desc("Slot to read item from."))
            .argument(MultiplePlayerSelectorArgument.optional("targets"), desc("Players to get items from."))
            .permission(perm("composite.mob"))
            .handler { handle(it, ::compositeReadItem) })
    }

    fun stats(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        fun sendTimings(baseTimings: Timings, headerKey: String) {
            plugin.sendMessage(sender, i18n.csafe(headerKey) {
                icu("time_last", baseTimings.allEntries().lastOrNull() ?: "?")
            })

            TIMING_INTERVALS.forEach { interval ->
                val intervalMs = interval * 1000L
                val timings = baseTimings.lastEntries(intervalMs)
                plugin.sendMessage(sender, i18n.csafe("stats.timing") {
                    icu("interval_ms", intervalMs)
                    icu("interval_sec", intervalMs / 1000.0)
                    icu("time_avg", timings.average())
                    icu("time_min", timings.min())
                    icu("time_max", timings.max())
                })
            }
        }

        sendTimings(plugin.engineTimings, "stats.engine_timings")

        plugin.sendMessage(sender, i18n.csafe("stats.object_types"))

        plugin.resolver.lastStats.forEach { (type, stats) ->
            val (candidates, updated) = stats
            val percent = updated.toDouble() / candidates
            plugin.sendMessage(sender, i18n.csafe("stats.object_type") {
                icu("type", type.key)
                icu("candidates", candidates)
                icu("updated", updated)
                icu("percent", percent)
            })
        }
    }

    fun systems(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val systemTypes = plugin.engine.systems()
            .mapNotNull { it::class.simpleName }

        plugin.sendMessage(sender, i18n.csafe("systems.header"))
        systemTypes.forEachIndexed { idx, type ->
            plugin.sendMessage(sender, i18n.csafe("systems.line") {
                icu("index", idx + 1)
                icu("type", type)
            })
        }
    }

    fun give(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val targets = ctx.players("targets", sender, i18n)
        val amount = ctx.value("amount") { 1 }
        val blueprint = ctx.get<EntityBlueprint>("blueprint")

        if (!plugin.hoster.canHost(blueprint, SokolObjectType.Item))
            error(i18n.safe("error.cannot_host"))

        targets.forEach { target ->
            val newBlueprint = blueprint.copyOf()
            newBlueprint.components.set(ItemHolder.byMob(target))
            val item = plugin.hoster.hostItem(newBlueprint)
            item.amount = amount
            target.inventory.addItem(item)
        }

        if (sender is Player)
            blueprint.components.set(ItemHolder.byMob(sender))
        val item = plugin.hoster.hostItem(blueprint)

        plugin.sendMessage(sender, i18n.csafe("give") {
            subst("item", item.displayName())
            subst("target", if (targets.size == 1) targets.first().name() else text(targets.size))
            icu("amount", amount)
        })
    }

    fun summon(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val location = ctx.get<Location>("location")
        val amount = ctx.value("amount") { 1 }
        val blueprint = ctx.get<KeyedEntityBlueprint>("blueprint")

        if (!plugin.hoster.canHost(blueprint, SokolObjectType.Mob))
            error(i18n.safe("error.cannot_host"))

        repeat(amount) {
            plugin.hoster.hostMob(blueprint, location)
        }

        plugin.sendMessage(sender, i18n.csafe("summon") {
            subst("id", text(blueprint.profile.id))
            icu("amount", amount)
        })
    }

    fun holdingFreeze(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val player = sender as Player
        val holding = player.alexandria.heldEntity ?: error(i18n.safe("error.not_holding"))
        val enable = ctx.value("enable") { !holding.frozen }

        holding.frozen = enable
        plugin.sendMessage(sender, i18n.csafe("holding.freeze.${if (enable) "enable" else "disable"}"))
    }

    fun holdingShape(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val player = sender as Player
        val holding = player.alexandria.heldEntity ?: error(i18n.safe("error.not_holding"))
        val shape = try {
            var node = ctx.get<ConfigurationNode>("config")
            // hocon serializers only accepts map nodes as root
            // if someone wants to define a compound `[...]` they use `{_:[...]}`
            if (node.hasChild("_"))
                node = node.node("_")
            node.force<Shape>()
        } catch (ex: SerializationException) {
            error(i18n.safe("error.parse_shape"), ex)
        }

        CraftBulletAPI.executePhysics {
            holding.drawShape = collisionOf(shape)
        }
    }

    fun holdingSlotShapes(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val player = sender as Player
        val holding = player.alexandria.heldEntity ?: error(i18n.safe("error.not_holding"))
        val enable = ctx.value("enable") { !holding.drawSlotShapes }

        holding.drawSlotShapes = enable
        plugin.sendMessage(sender, i18n.csafe("holding.slot_shapes.${if (enable) "enable" else "disable"}"))
    }

    fun stateRead(
        ctx: Context,
        sender: CommandSender,
        i18n: I18N<Component>,
        entity: SokolEntity,
        name: Component,
    ) {
        val components = entity.components.all()

        val componentTypes = components
            .mapNotNull { it.componentType.simpleName }
            .sorted()

        val hover = (
            i18n.csafe("state.read.component.header") +
            componentTypes.flatMap { type ->
                i18n.csafe("state.read.component.line") {
                    icu("type", type)
                }
            }
        ).join(JoinConfiguration.newlines())

        val node = AlexandriaAPI.configLoader().build().createNode()
        node.set(entity)

        val render = node.render()
        plugin.sendMessage(sender, i18n.csafe("state.read.header.${if (render.isEmpty()) "empty" else "present"}") {
            subst("entity", name)
        }.map { it.hoverEvent(hover) })

        plugin.sendMessage(sender, render.flatMap { line ->
            i18n.csafe("state.read.line") {
                subst("line", line)
            }
        })
    }

    fun stateReadMob(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val targets = ctx.entities("targets", sender, i18n)

        var results = 0
        targets.forEach { target ->
            plugin.useMob(target, false) { entity ->
                stateRead(ctx, sender, i18n, entity, target.name())
                results++
            }
        }

        plugin.sendMessage(sender, i18n.csafe("state.read.complete") {
            icu("results", results)
        })
    }

    fun stateReadItem(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val slot = ctx.value<PlayerInventorySlot>("slot") { PlayerInventorySlot.ByEquipment(EquipmentSlot.HAND) }
        val targets = ctx.players("targets", sender, i18n)

        var results = 0
        targets.forEach { target ->
            val item = slot.getFrom(target.inventory)
            plugin.useItem(item, false,
                { blueprint ->
                    blueprint.components.set(ItemHolder.byPlayer(target, slot.asInt(target.inventory)))
                }
            ) { entity ->
                stateRead(ctx, sender, i18n, entity, item.displayName())
                results++
            }
        }

        plugin.sendMessage(sender, i18n.csafe("state.read.complete") {
            icu("results", results)
        })
    }

    fun stateWriteMob(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        val targets = ctx.entities("targets", sender, i18n)
        val data = ctx.get<ConfigurationNode>("data")

        /*
        val dataMap = try {
            data.forceMap(SokolBlueprint::class.java).map { (_, child) ->
                val type = componentTypeFrom(plugin, SokolBlueprint::class.java, child)
                type to child
            }.associate { it }
        } catch (ex: SerializationException) {
            error(i18n.safe("error.parse_blueprint"), ex)
        }

        var results = 0
        targets.forEach { target ->
            plugin.useMob(target) { entity ->
                dataMap.forEach { (type, child) ->
                    val engineType = plugin.engine.componentType(type.componentType)
                    entity.getComponent(engineType)?.let { existing ->
                        if (existing is PersistentComponent) {
                            // merge cfg values
                            val configNode = AlexandriaAPI.configLoader().build().createNode()
                            existing.write(configNode)
                            child.mergeFrom(configNode)

                            val newComponent = try {
                                type.read(child)
                            } catch (ex: SerializationException) {
                                error(i18n.safe("error.parse_blueprint"), ex)
                            }

                            entity.setComponent(newComponent)
                        }
                    } ?: run {
                        // add new component
                        val newComponent = try {
                            type.read(child)
                        } catch (ex: SerializationException) {
                            error(i18n.safe("error.parse_blueprint"), ex)
                        }
                        entity.setComponent(newComponent)
                    }
                }
                results++
            }
        }

        plugin.sendMessage(sender, i18n.csafe("state.write.complete") {
            icu("results", results)
        })*/
    }

    fun compositeReadMob(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        // todo
    }

    fun compositeReadItem(ctx: Context, sender: CommandSender, i18n: I18N<Component>) {
        // todo
    }*/
}
